nodes/curve/marching_squares_on_surface.py

```diff
@@ -113,8 +113,8 @@ def process(self):
 
                 contours = measure.find_contours(field_values, level=value)
 
-                u_size = (u_max - u_min)/samples_u
-                v_size = (v_max - v_min)/samples_v
+                u_size = (u_max - u_min)/(samples_u-1)-0.0000001
+                v_size = (v_max - v_min)/(samples_v-1)-0.0000001
```

Comment:
I would suggest to make this a node setting, instead of hardcoding. Maybe hide it in node's N panel.

Reply:
@portnov This is not good idea. I found real problem and wrote new fix. The description of problem:

With some T value calculation do errors:
![image](https://user-images.githubusercontent.com/14288520/208514870-cd69142c-e564-4097-8898-56933201a828.png)

The problem is (4):

u_size = **(u_max - u_min)/(samples_u-1)**
**u_size * (samples_u-1) > (u_max - u_min)** (WTF??? - float precision)

and this go to next wrong result:
![image](https://user-images.githubusercontent.com/14288520/208515920-190165d4-b805-4efa-bef5-e9fdca43e23b.png)

In this situation need decrease a bit **u_size** and I propose do this with next code:
![image](https://user-images.githubusercontent.com/14288520/208516418-160c468c-ae34-40c0-bcb8-91bd05f561d4.png)

This take good result:
![image](https://user-images.githubusercontent.com/14288520/208516480-539d7cff-1594-4993-a410-c4dafdf2070d.png)

If this is not enough then this can be do with (for range) **u_size/v_size** to calc a value where
**u_size * (samples_u-1) <= (u_max - u_min)**

What do you think?

Reply:
![MarchingSquaresOnSurface](https://user-images.githubusercontent.com/14288520/208519426-16a62404-135a-4e86-9192-f6653c7224d9.gif)

Reply:
Found some glitches. I create for-range with new formula based on:

https://github.com/nortikin/sverchok/blob/b7b06ebc9b61f0002463973f3648dd30f8a807db/utils/marching_squares.py#L31-L32

https://gist.github.com/df4d5b173e4934aff740746f28450cf1
![MarchingSquaresOnSurface fix 002](https://user-images.githubusercontent.com/14288520/208533991-5b9efa72-bc44-473d-a0be-9f2a28126745.gif)

Reply:
There is a connection in current version between these lines of code:

https://github.com/nortikin/sverchok/blob/b7b06ebc9b61f0002463973f3648dd30f8a807db/nodes/curve/marching_squares_on_surface.py#L116-L119

https://github.com/nortikin/sverchok/blob/b7b06ebc9b61f0002463973f3648dd30f8a807db/utils/marching_squares.py#L31-L32

**x_size = u_size** and on last iteration in line 31 result of **x** has to be **u_max** but with some **T** values result exceed **u_max**. This is a main problem. If refactor of **make_contours** function and send **u_max** and **v_max** as addition parameters and assign its in last [for-range] cycle like:

x = min_x + x_size * x0
y = min_y + y_size * y0

if x0==samples_u-1: x=u_max

if y0==samples_u-1: y=v_max

then problem will solved forever )))

Reply:
you can always add new parameter to make_contours if you make it optional.

Reply:
@portnov Thank you! It work. Can you look at that?

Reply:
https://gist.github.com/faa14b885442ea6a6388e6bf103faeb5
![Curves-MarchinSquares v004 edited](https://user-images.githubusercontent.com/14288520/208666524-b8e9bf04-5e08-42e3-aabf-a307dee468d6.gif)